Subtest `repository_files` in IT fails with an assertion error

https://gitlab.prod.anvil.gi.ucsc.edu/ucsc/azul/-/jobs/17859/raw

```
…
2023-12-14 17:27:00,726    INFO MainThread test.integration_test: Beginning sub-test [repository_files] {'catalog': 'anvil3-it'}
2023-12-14 17:27:00,727    INFO MainThread test.integration_test: Making GET request to 'https://service.prod.anvil.gi.ucsc.edu/fetch/repository/files/f6c7670c-2fa2-48eb-aa74-31b7ec94b148?catalog=anvil3-it&version=2022-06-01T00%3A00%3A00.000000Z'
2023-12-14 17:27:00,727   DEBUG MainThread test.integration_test: … without request body
2023-12-14 17:27:13,203    INFO MainThread test.integration_test: Got 200 response after 12.476s from GET to https://service.prod.anvil.gi.ucsc.edu/fetch/repository/files/f6c7670c-2fa2-48eb-aa74-31b7ec94b148?catalog=anvil3-it&version=2022-06-01T00%3A00%3A00.000000Z
2023-12-14 17:27:13,204   DEBUG MainThread test.integration_test: … with response headers HTTPHeaderDict({'Content-Type': 'application/json', 'Content-Length': '1035', 'Connection': 'keep-alive', 'Date': 'Thu, 14 Dec 2023 17:27:13 GMT', 'x-amzn-RequestId': '78106650-10f0-4a15-8922-dde51ff0efed', 'Access-Control-Allow-Origin': '*', 'Strict-Transport-Security': 'max-age=31536000; includeSubDomains', 'Access-Control-Allow-Headers': 'Authorization,Content-Type,X-Amz-Date,X-Amz-Security-Token,X-Api-Key', 'x-amz-apigw-id': 'P8YlcGtKoAMEopw=', 'X-Amzn-Trace-Id': 'Root=1-657b3aef-178257e03a3343e74146d6dd;Sampled=0;lineage=45061563:0', 'X-Cache': 'Miss from cloudfront', 'Via': '1.1 51391527dd8c879c45b44b119905c872.cloudfront.net (CloudFront)', 'X-Amz-Cf-Pop': 'IAD89-C3', 'X-Amz-Cf-Id': 'Ul50UYg5UQISCs_BcDu-t2Apu8WBuBU0_T4qN12bINo-SrSNKWIEqg=='})
2023-12-14 17:27:13,204   DEBUG MainThread test.integration_test: … with response body b'{"Status": 302, "Location": "https://storage.googleapis.com/fc-47de7dae-e8e6-429c-b760-b4ba49136eee/summary_stats/variant_cou...'
2023-12-14 17:27:13,205    INFO MainThread test.integration_test: Making GET request to 'https://storage.googleapis.com/fc-47de7dae-e8e6-429c-b760-b4ba49136eee/summary_stats/variant_counts_windows/dchr20.GIH.counts.txt?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=tdr-ingest-sa%40datarepo-9e80499e.iam.gserviceaccount.com%2F20231214%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20231214T172713Z&X-Goog-Expires=900&X-Goog-SignedHeaders=host&requestedBy=azul-ucsc-1-public-anvilprod%40platform-anvil-prod.iam.gserviceaccount.com&userProject=datarepo-02ad84ea&X-Goog-Signature=592e011867cdc11a940295919de1c105f6bc43230b1b43bac70d84572642603479a24dde23fc41497261d12527b4c371533334279dbba6c31b3a7a0ff865e3f9ec352f0156c12cb592eb72058991d0e02d7997a9119eb3acd937711f9cda6b5f9beb6638fd25741e677572bcb3a08822e1889ccf2329a4d4ac94e7cdae0f1f234d9caf72f35817e952b277484c05e05d39462a207631f1d90fe7b283758785136c671a2c7aedcb50563a2cc0c6c608cb461ae60dd7e40a5cda690326496f5b42dea5d24d815ee495ccaf8a0281ee3feda43a0846d8173196f86a8063a8a0e7a67068c689dbb09859f4fd7bdb8f2859e08d43342c8128b1d4d003208c0fffa0d7'
2023-12-14 17:27:13,205   DEBUG MainThread test.integration_test: … without request body
2023-12-14 17:27:13,395    INFO MainThread test.integration_test: Got 200 response after 0.189s from GET to https://storage.googleapis.com/fc-47de7dae-e8e6-429c-b760-b4ba49136eee/summary_stats/variant_counts_windows/dchr20.GIH.counts.txt?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=tdr-ingest-sa%40datarepo-9e80499e.iam.gserviceaccount.com%2F20231214%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20231214T172713Z&X-Goog-Expires=900&X-Goog-SignedHeaders=host&requestedBy=azul-ucsc-1-public-anvilprod%40platform-anvil-prod.iam.gserviceaccount.com&userProject=datarepo-02ad84ea&X-Goog-Signature=592e011867cdc11a940295919de1c105f6bc43230b1b43bac70d84572642603479a24dde23fc41497261d12527b4c371533334279dbba6c31b3a7a0ff865e3f9ec352f0156c12cb592eb72058991d0e02d7997a9119eb3acd937711f9cda6b5f9beb6638fd25741e677572bcb3a08822e1889ccf2329a4d4ac94e7cdae0f1f234d9caf72f35817e952b277484c05e05d39462a207631f1d90fe7b283758785136c671a2c7aedcb50563a2cc0c6c608cb461ae60dd7e40a5cda690326496f5b42dea5d24d815ee495ccaf8a0281ee3feda43a0846d8173196f86a8063a8a0e7a67068c689dbb09859f4fd7bdb8f2859e08d43342c8128b1d4d003208c0fffa0d7
2023-12-14 17:27:13,395   DEBUG MainThread test.integration_test: … with response headers HTTPHeaderDict({'X-GUploader-UploadID': 'ABPtcPr97tilFR4kHP9QPnt9W-chNXbW_NWLgyBwMEtumNFUcd91ymzYlh4OWWuApF3ZYlOSO34vA8w2Pw', 'Last-Modified': 'Fri, 26 Aug 2022 16:31:18 GMT', 'ETag': '"d41d8cd98f00b204e9800998ecf8427e"', 'x-goog-generation': '1661531478218204', 'x-goog-metageneration': '1', 'x-goog-stored-content-encoding': 'identity', 'x-goog-stored-content-length': '0', 'x-goog-meta-x-goog-reserved-source-generation': '1661530046474186', 'Content-Type': 'text/plain', 'x-goog-hash': 'crc32c=AAAAAA==, md5=1B2M2Y8AsgTpgAmY7PhCfg==', 'x-goog-storage-class': 'STANDARD', 'Accept-Ranges': 'bytes', 'Content-Length': '0', 'Cache-Control': 'no-cache, no-store, max-age=0, must-revalidate', 'Pragma': 'no-cache', 'Expires': 'Mon, 01 Jan 1990 00:00:00 GMT', 'Date': 'Thu, 14 Dec 2023 17:27:13 GMT', 'Server': 'UploadServer', 'Alt-Svc': 'h3=":443"; ma=2592000,h3-29=":443"; ma=2592000'})
2023-12-14 17:27:13,395   DEBUG MainThread test.integration_test: … with a streamed response body
2023-12-14 17:27:13,395    INFO MainThread test.integration_test: Failed sub-test [repository_files] {'catalog': 'anvil3-it'}

  test_indexing (integration_test.IndexingIntegrationTest.test_indexing) [repository_files] (catalog='anvil3-it') ... FAIL
…
======================================================================
FAIL: test_indexing (integration_test.IndexingIntegrationTest.test_indexing) [repository_files] (catalog='anvil3-it')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builds/ucsc/azul/test/integration_test.py", line 411, in subTest
    yield
  File "/builds/ucsc/azul/test/integration_test.py", line 988, in _test_repository_files
    self._validate_file_response(response, self._file_ext(file))
  File "/builds/ucsc/azul/test/integration_test.py", line 1014, in _validate_file_response
    self._validate_file_content(response, file_ext)
  File "/builds/ucsc/azul/test/integration_test.py", line 1005, in _validate_file_content
    self.assertEqual(1, len(content.read(1)))
AssertionError: 1 != 0
----------------------------------------------------------------------
Ran 15 tests in 1346.798s
FAILED (failures=1, skipped=1)
make: *** [Makefile:241: integration_test] Error 1
Cleaning up project directory and file based variables 00:01
ERROR: Job failed: exit code 1
```
<!--
Add your description here. For suspected bugs, the steps required to reproduce
the problem, as well as the expected and actual outcome should be included in
the description. Solutions should be proposed in a  separate comment, not the
issue description. Please do not modify or delete the checklist below. The
checklist should remain at the end of the issue description.
-->

- [x] Security design review completed; the Resolution of this issue does **not** …
  - [x] … affect authentication; for example:
    - OAuth 2.0 with the application (API or Swagger UI)
    - Authentication of developers with Google Cloud APIs
    - Authentication of developers with AWS APIs
    - Authentication with a GitLab instance in the system
    - Password and 2FA authentication with GitHub
    - API access token authentication with GitHub
    - Authentication with 
  - [x] … affect the permissions of internal users like access to
    - Cloud resources on AWS and GCP
    - GitLab repositories, projects and groups, administration
    - an EC2 instance via SSH
    - GitHub issues, pull requests, commits, commit statuses, wikis, repositories, organizations
  - [x] … affect the permissions of external users like access to
    - TDR snapshots
  - [x] … affect permissions of service or bot accounts
    - Cloud resources on AWS and GCP
  - [x] … affect audit logging in the system, like
    - adding, removing or changing a log message that represents an auditable event
    - changing the routing of log messages through the system
  - [x] … affect monitoring of the system
  - [x] … introduce a new software dependency like
    - Python packages on PYPI
    - Command-line utilities
    - Docker images
    - Terraform providers
  - [x] … add an interface that exposes sensitive or confidential data at the security boundary
  - [x] … affect the encryption of data at rest
  - [x] … require persistence of sensitive or confidential data that might require encryption at rest
  - [x] … require unencrypted transmission of data within the security boundary
  - [x] … affect the network security layer; for example by 
    - modifying, adding or removing firewall rules
    - modifying, adding or removing security groups
    - changing or adding a port a service, proxy or load balancer listens on
- [x] Documentation on any unchecked boxes is provided in comments below

--- a/test/integration_test.py
+++ b/test/integration_test.py
@@ -43,6 +43,7 @@ from typing import (
     IO,
     Optional,
     Protocol,
+    TypedDict,
     cast,
 )
 import unittest
@@ -198,6 +199,13 @@ class ReadableFileObject(Protocol):
     def read(self, amount: int) -> bytes: ...
 
     def seek(self, amount: int) -> Any: ...
+
+
+class FileInnerEntity(TypedDict):
+    uuid: str
+    version: str
+    name: str
+    size: int
 
 
 GET = 'GET'
@@ -645,10 +653,10 @@ class IndexingIntegrationTest(IntegrationTestCase, AlwaysTearDownTestCase):
         execution_ids = {token.execution_id for token in tokens}
         return execution_ids
 
-    def _get_one_inner_file(self, catalog: CatalogName) -> JSON:
+    def _get_one_inner_file(self, catalog: CatalogName) -> FileInnerEntity:
         outer_file = self._get_one_outer_file(catalog)
         inner_files: JSONs = outer_file['files']
-        return one(inner_files)
+        return cast(FileInnerEntity, one(inner_files))
 
     @cache
     def _get_one_outer_file(self, catalog: CatalogName) -> JSON:
@@ -708,9 +716,8 @@ class IndexingIntegrationTest(IntegrationTestCase, AlwaysTearDownTestCase):
     def _test_dos_and_drs(self, catalog: CatalogName):
         if config.is_dss_enabled(catalog) and config.dss_direct_access:
             file = self._get_one_inner_file(catalog)
-            file_uuid, file_ext = file['uuid'], self._file_ext(file)
-            self._test_dos(catalog, file_uuid, file_ext)
-            self._test_drs(catalog, file_uuid, file_ext)
+            self._test_dos(catalog, file)
+            self._test_drs(catalog, file)
 
     @property
     def _service_account_credentials(self) -> ContextManager:
@@ -985,9 +992,9 @@ class IndexingIntegrationTest(IntegrationTestCase, AlwaysTearDownTestCase):
                     response = self._get_url_json(GET, furl(response['Location']))
                 self.assertNotIn('Retry-After', response)
                 response = self._get_url(GET, furl(response['Location']), stream=True)
-                self._validate_file_response(response, self._file_ext(file))
+                self._validate_file_response(response, file)
 
-    def _file_ext(self, file: JSON) -> str:
+    def _file_ext(self, file: FileInnerEntity) -> str:
         # We believe that the file extension is a more reliable indicator than
         # the `format` metadata field. Note that this method preserves multipart
         # extensions and includes the leading '.', so the extension of
@@ -995,49 +1002,50 @@ class IndexingIntegrationTest(IntegrationTestCase, AlwaysTearDownTestCase):
         suffixes = PurePath(file['name']).suffixes
         return ''.join(suffixes).lower()
 
-    def _validate_file_content(self, content: ReadableFileObject, file_ext: str):
+    def _validate_file_content(self, content: ReadableFileObject, file: FileInnerEntity):
+        file_ext = self._file_ext(file)
         if file_ext == '.fastq':
             self._validate_fastq_content(content)
         elif file_ext == '.fastq.gz':
             with gzip.open(content) as buf:
                 self._validate_fastq_content(buf)
         else:
-            self.assertEqual(1, len(content.read(1)))
+            self.assertEqual(1 if file['size'] > 0 else 0, len(content.read(1)))
 
     def _validate_file_response(self,
                                 response: urllib3.HTTPResponse,
-                                file_ext: str):
+                                file: FileInnerEntity):
         """
         Note: The response object must have been obtained with stream=True
         """
         try:
-            self._validate_file_content(response, file_ext)
+            self._validate_file_content(response, file)
         finally:
             response.close()
 
-    def _test_drs(self, catalog: CatalogName, file_uuid: str, file_ext: str):
+    def _test_drs(self, catalog: CatalogName, file: FileInnerEntity):
         repository_plugin = self.azul_client.repository_plugin(catalog)
         drs = repository_plugin.drs_client()
         for access_method in AccessMethod:
             with self.subTest('drs', catalog=catalog, access_method=AccessMethod.https):
-                log.info('Resolving file %r with DRS using %r', file_uuid, access_method)
-                drs_uri = f'drs://{config.api_lambda_domain("service")}/{file_uuid}'
+                log.info('Resolving file %r with DRS using %r', file['uuid'], access_method)
+                drs_uri = f'drs://{config.api_lambda_domain("service")}/{file["uuid"]}'
                 access = drs.get_object(drs_uri, access_method=access_method)
                 self.assertIsNone(access.headers)
                 if access.method is AccessMethod.https:
                     response = self._get_url(GET, furl(access.url), stream=True)
-                    self._validate_file_response(response, file_ext)
+                    self._validate_file_response(response, file)
                 elif access.method is AccessMethod.gs:
                     content = self._get_gs_url_content(furl(access.url), size=self.num_fastq_bytes)
-                    self._validate_file_content(content, file_ext)
+                    self._validate_file_content(content, file)
                 else:
                     self.fail(access_method)
 
-    def _test_dos(self, catalog: CatalogName, file_uuid: str, file_ext: str):
+    def _test_dos(self, catalog: CatalogName, file: FileInnerEntity):
         with self.subTest('dos', catalog=catalog):
-            log.info('Resolving file %s with DOS', file_uuid)
+            log.info('Resolving file %s with DOS', file['uuid'])
             response = self._check_endpoint(method=GET,
-                                            path=drs.dos_object_url_path(file_uuid),
+                                            path=drs.dos_object_url_path(file['uuid']),
                                             args=dict(catalog=catalog))
             json_data = json.loads(response)['data_object']
             file_url = first(json_data['urls'])['url']
@@ -1057,7 +1065,7 @@ class IndexingIntegrationTest(IntegrationTestCase, AlwaysTearDownTestCase):
                     else:
                         break
             self._assertResponseStatus(response, 200)
-            self._validate_file_content(response, file_ext)
+            self._validate_file_content(response, file)
 
     def _get_gs_url_content(self,
                             url: furl,

